astropy/units/equivalencies.py

```diff
@@ -205,145 +215,34 @@ def spectral_density(wav, factor=None):
-    phot_L_nu = astrophys.photon / (si.s * si.Hz)
-
-    # surface brightness (flux equiv)
-    S_la = cgs.erg / si.angstrom / si.cm**2 / si.s / si.sr
-    S_nu = cgs.erg / si.Hz / si.cm**2 / si.s / si.sr
-    nu_S_nu = cgs.erg / si.cm**2 / si.s / si.sr
-    la_S_la = nu_S_nu
-    phot_S_la = astrophys.photon / (si.cm**2 * si.s * si.AA * si.sr)
-    phot_S_nu = astrophys.photon / (si.cm**2 * si.s * si.Hz * si.sr)
-
-    # surface brightness (luminosity equiv)
-    SL_nu = cgs.erg / si.s / si.Hz / si.sr
-    SL_la = cgs.erg / si.s / si.angstrom / si.sr
-    nu_SL_nu = cgs.erg / si.s / si.sr
-    la_SL_la = nu_SL_nu
-    phot_SL_la = astrophys.photon / (si.s * si.AA * si.sr)
-    phot_SL_nu = astrophys.photon / (si.s * si.Hz * si.sr)
-
-    def f_la_to_f_nu(x):
-        return x * (wav.to_value(si.AA, spectral()) ** 2 / c_Aps)
-
-    def f_la_from_f_nu(x):
-        return x / (wav.to_value(si.AA, spectral()) ** 2 / c_Aps)
-
-    def f_nu_to_nu_f_nu(x):
-        return x * wav.to_value(si.Hz, spectral())
-
-    def f_nu_from_nu_f_nu(x):
-        return x / wav.to_value(si.Hz, spectral())
-
-    def f_la_to_la_f_la(x):
-        return x * wav.to_value(si.AA, spectral())
-
-    def f_la_from_la_f_la(x):
-        return x / wav.to_value(si.AA, spectral())
-
-    def phot_f_la_to_f_la(x):
-        return hc * x / wav.to_value(si.AA, spectral())
-
-    def phot_f_la_from_f_la(x):
-        return x * wav.to_value(si.AA, spectral()) / hc
-
-    def phot_f_la_to_f_nu(x):
-        return h_cgs * x * wav.to_value(si.AA, spectral())
-
-    def phot_f_la_from_f_nu(x):
-        return x / (wav.to_value(si.AA, spectral()) * h_cgs)
-
-    def phot_f_la_to_phot_f_nu(x):
-        return x * wav.to_value(si.AA, spectral()) ** 2 / c_Aps
-
-    def phot_f_la_from_phot_f_nu(x):
-        return c_Aps * x / wav.to_value(si.AA, spectral()) ** 2
-
-    phot_f_nu_to_f_nu = phot_f_la_to_f_la
-    phot_f_nu_from_f_nu = phot_f_la_from_f_la
-
-    def phot_f_nu_to_f_la(x):
-        return x * hc * c_Aps / wav.to_value(si.AA, spectral()) ** 3
-
-    def phot_f_nu_from_f_la(x):
-        return x * wav.to_value(si.AA, spectral()) ** 3 / (hc * c_Aps)
-
-    # for luminosity density
-    L_nu_to_nu_L_nu = f_nu_to_nu_f_nu
-    L_nu_from_nu_L_nu = f_nu_from_nu_f_nu
-    L_la_to_la_L_la = f_la_to_la_f_la
-    L_la_from_la_L_la = f_la_from_la_f_la
-
-    phot_L_la_to_L_la = phot_f_la_to_f_la
-    phot_L_la_from_L_la = phot_f_la_from_f_la
-    phot_L_la_to_L_nu = phot_f_la_to_f_nu
-    phot_L_la_from_L_nu = phot_f_la_from_f_nu
-    phot_L_la_to_phot_L_nu = phot_f_la_to_phot_f_nu
-    phot_L_la_from_phot_L_nu = phot_f_la_from_phot_f_nu
-    phot_L_nu_to_L_nu = phot_f_nu_to_f_nu
-    phot_L_nu_from_L_nu = phot_f_nu_from_f_nu
-    phot_L_nu_to_L_la = phot_f_nu_to_f_la
-    phot_L_nu_from_L_la = phot_f_nu_from_f_la
+
+    wav_in_AA = wav.to_value(si.AA, equivalencies=spectral())
+
+    def func_factory(factor):
+        return lambda x: x * factor, lambda x: x * factor**-1
```

Comment:
This helper shortens the code a lot, but it also forces pre-computing the conversion factors, whereas the previous code converted the factors lazily. If the `units` maintainers are not happy with the performance of the updated code then it might be best to reintroduce the explicit function definitions.

Reply:
If the func_factories are replaced with their explicit counterparts, would that be sufficient for the lazy computation?

Reply:
Yes, but the tradeoff is that the code would be much more verbose.

Reply:
Would it be 14 lines (7 func_factory x 2 lambda funcs per factory)?